app/src/main/java/com/example/wrk/fragments/FeedFragment.java

```diff
@@ -58,30 +70,122 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
+        performedQuery.addDescendingOrder(WorkoutPerformed.KEY_CREATED_AT);
+        performedQuery.setLimit(5);
+        // specify workouts
+        performedQuery.whereContainedIn(WorkoutPerformed.KEY_USER, ParseUser.getCurrentUser().getList("following"));
+
+        List<WorkoutPerformed> list = new ArrayList<>();
+        recentFollowedWorkouts = new ArrayList<>();
+        list.addAll(performedQuery.find());
+        for (int i = 0; i < list.size(); i++) {
+            LocalDate postDate = list.get(i)
+                    .getCreatedAt().toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
+            // only add the posts within the past 2 days
+            if (postDate.isAfter(daysAgoDate)) {
+                recentFollowedWorkouts.add(list.get(i));
+            }
+        }
+        workoutsPerformed.addAll(recentFollowedWorkouts);
+    }
+
+    protected void queryPopularWorkouts() throws ParseException {
+        ParseQuery<WorkoutTemplate> performedQuery = ParseQuery.getQuery(WorkoutTemplate.class);
+        // includes specified data
+        performedQuery.include(WorkoutTemplate.KEY_TITLE);
+        performedQuery.include(WorkoutTemplate.KEY_COMPONENTS);
+        performedQuery.addDescendingOrder(WorkoutTemplate.KEY_SAVED_BY);
+        List<WorkoutTemplate> allTemplates = new ArrayList<>();
+        allTemplates.addAll(performedQuery.find());
+
+        popularTemplates = new ArrayList<>();
+        mostPopularWorkouts = new ArrayList<>();
+
+        // compare amount of users that saved the templates
+        Comparator<WorkoutTemplate> workoutTemplateComparator = new Comparator<WorkoutTemplate>() {
+            @Override
+            public int compare(WorkoutTemplate t1, WorkoutTemplate t2) {
+                return Integer.compare(t1.getSavedBy().size(), t2.getSavedBy().size());
+            }
+        };
+        // sort templates by greatest to least in save count
+        Collections.sort(allTemplates, workoutTemplateComparator);
+        Collections.reverse(allTemplates);
+        // place up to top 3 templates into the popular ones
+        for (int i = 0; i < allTemplates.size(); i++) {
+            if (i == 3) {
+                break;
+            }
+            else {
+                popularTemplates.add(allTemplates.get(i));
+            }
+        }
+        ParseQuery<WorkoutPerformed> parseQuery = ParseQuery.getQuery(WorkoutPerformed.class);
+        parseQuery.include(WorkoutPerformed.KEY_USER);
+        parseQuery.include(WorkoutPerformed.KEY_WORKOUT);
+        // look for posts that are within the popular templates
+        parseQuery.whereContainedIn(WorkoutPerformed.KEY_WORKOUT, popularTemplates);
+        // newest first
+        parseQuery.addDescendingOrder(WorkoutPerformed.KEY_CREATED_AT);
+        mostPopularWorkouts.addAll(parseQuery.find());
+
+        // remove if template is the same as another post
+        for (int i = 0; i < mostPopularWorkouts.size(); i++) {
```

Comment:
Ah I understand the problem you tried to solve, but the logic here is really error-prone since you're iterating on the list while modifying the list at the same time. 

I think the better way to do it is checking if it's duplicated workout when adding to mostPopularWorkouts. If it's duplicated, just don't add it

Reply:
> Ah I understand the problem you tried to solve, but the logic here is really error-prone since you're iterating on the list while modifying the list at the same time.
> 
> I think the better way to do it is checking if it's duplicated workout when adding to mostPopularWorkouts. If it's duplicated, just don't add it

I did consider that but I ran into the issue that the query doesn't add one at a time, it fetches them all and returns a list of the found results. Without an existing list of results to filter by, I can't seem to figure out a way to implement that without making another query first of all the posts first and then doing it. This just seems like it would slow down the loading as it would have to make another query. Also, the queries are for posts which each have a pointer to a template so I am not sure a way to check for a duplicate template without gathering all of the posts first and then sorting to check each one for only the most recent template. Any suggestions?

Reply:
Have you tried using parseQuery.each()? Seems can iterate the query results one by one
Maybe something like 
```
Set<Integer> existedWorkoutIds = new Hashset<>();
parseQuery.each((workout) => if(existedWorkoutIds.!contains(workout.getId(){
mostPopularWorkouts.add(workout);
workoutIdSet.add(workout.getId());
}));
```
Not sure if this will work though, feel free to move on if it's not

Reply:
Good jobs on your findings! Yeah I think that API doesn't exist in Java so please ignore my comment above.

I think it's good to keep this logic for now since we want to both deduplicate and keeping the sorting order at the same time. Maybe some more complicated data structure can do that, but don't think worthwhile

Reply:
Okay thanks!